examples/algorithm_progression.py

```diff
@@ -0,0 +1,172 @@
+argparser.add_argument(
+    '-o',
+    '--output_dir',
+    required=False,
+    help='Name for a new directory in this location, wherein result images will be stored',
+)
+argparser.add_argument(
+    '-w',
+    '--num_workers',
+    required=False,
+    default=4,
+    type=int,
+    help='Number of workers to use for processing alogrithm iterations',
+)
+argparser.add_argument(
+    '-p',
+    '--param',
+    action='append',
+    required=False,
+    nargs='*',
+    help='A parameter to pass to the algorithm; '
+         'instead of using the default value, '
+         'Pass every item in the number space, '
+         'specified as `--param=param_name,start,end,num_items[,open]`',
+)
+
+
+def apply_algorithm(algorithm, input_filename, output_dir, params, unique_params):
+    source = large_image.open(input_filename)
+    filename = f'{algorithm.__name__}_{"_".join(["{:.2f}".format(v) for v in unique_params])}.tiff'
+    filepath = Path(output_dir, filename)
+    desc = {'path': filename}
+    for i, v in enumerate(unique_params):
+        desc[VARIABLE_LAYERS[i]] = float(v)
+
+    new_source = large_image.new()
+    for tile in source.tileIterator(
+        format=large_image.tilesource.TILE_FORMAT_NUMPY,
+        tile_size=dict(width=2048, height=2048),
+    ):
+        altered_data = algorithm(tile['tile'], *params)
+    new_source.addTile(altered_data, tile['x'], tile['y'])
+    new_source.write(str(filepath), lossy=False)
+    return desc
+
+
+def sweep_algorithm(algorithm, input_filename, input_params, output_dir, max_workers):
+    algorithm_name = algorithm.__name__.replace('_', ' ').title()
+    yaml_dict = {
+        'name': f'{algorithm_name} iterative results',
+        'description': f'{algorithm_name} algorithm performed on {input_filename}',
+        'sources': [{'path': input_filename}],
+    }
+    param_desc = [
+        f'{VARIABLE_LAYERS[i]} = {n}'
+        for i, (n, p) in enumerate(input_params.items())
+        if len(p) > 1 and i < len(VARIABLE_LAYERS)
+    ]
+    if len(param_desc) > 0:
+        yaml_dict['description'] += f', where {", ".join(param_desc)}'
+
+    param_space_combos = list(itertools.product(*input_params.values()))
+    combos_uniques = list(
+        itertools.product(*[p for p in input_params.values() if len(p) > 1])
+    )
+
+    print(f'Beginning {len(param_space_combos)} runs on {max_workers} workers...')
+    with concurrent.futures.ProcessPoolExecutor(max_workers=max_workers) as executor:
+        # cannot pass source through submitted task; it is unpickleable
+        futures = [
+            executor.submit(
+                apply_algorithm,
+                algorithm,
+                input_filename,
+                output_dir,
+                combo,
+                unique,
+            )
+            for combo, unique in zip(param_space_combos, combos_uniques)
+        ]
+        for future in enumerate(concurrent.futures.as_completed(futures)):
```

Comment:
If you use `enumerate`, you are iterating over a tuple `(index, future)`, not just the values in the futures.

Reply:
Changed in [8f12e60](https://github.com/girder/large_image/pull/1051/commits/8f12e60fa621de791b9463118c999c13ddb2e88a)